src/ledger/database/lmdb.zig

```diff
@@ -0,0 +1,641 @@
+const std = @import("std");
+const c = @import("lmdb");
+const sig = @import("../../sig.zig");
+const database = @import("lib.zig");
+const build_options = @import("build-options");
+
+const Allocator = std.mem.Allocator;
+
+const BytesRef = database.interface.BytesRef;
+const ColumnFamily = database.interface.ColumnFamily;
+const IteratorDirection = database.interface.IteratorDirection;
+const Logger = sig.trace.Logger;
+
+const key_serializer = database.interface.key_serializer;
+const value_serializer = database.interface.value_serializer;
+
+pub fn LMDB(comptime column_families: []const ColumnFamily) type {
+    return struct {
+        allocator: Allocator,
+        env: *c.MDB_env,
+        dbis: []const c.MDB_dbi,
+        path: [:0]const u8,
+
+        const Self = @This();
+
+        pub fn open(allocator: Allocator, logger: Logger, path: []const u8) anyerror!Self {
+            logger.info().log("Initializing LMDB");
+            const owned_path = try allocator.dupeZ(u8, path);
+
+            // create and open the database
+            const env = try returnOutput(c.mdb_env_create, .{});
+            try maybeError(c.mdb_env_set_maxdbs(env, column_families.len));
+            try maybeError(c.mdb_env_open(env, owned_path.ptr, 0, 0o700));
+
+            // begin transaction to create column families aka "databases" in lmdb
+            const txn = try returnOutput(c.mdb_txn_begin, .{ env, null, 0 });
+            errdefer c.mdb_txn_abort(txn);
+
+            // allocate cf handles
+            const dbis = try allocator.alloc(c.MDB_dbi, column_families.len);
+            errdefer allocator.free(dbis);
+
+            // save cf handles
+            inline for (column_families, 0..) |cf, i| {
+                // open cf/database, creating if necessary
+                dbis[i] = try returnOutput(c.mdb_dbi_open, .{ txn, cf.name.ptr, c.MDB_CREATE });
+            }
+
+            // persist column families
+            try maybeError(c.mdb_txn_commit(txn));
+
+            return .{
+                .allocator = allocator,
+                .env = env,
+                .dbis = dbis,
+                .path = owned_path,
+            };
+        }
+
+        pub fn deinit(self: *Self) void {
+            self.allocator.free(self.dbis);
```

Comment:
More of a question. There is also `self.env`, which is a pointer. What takes care of freeing that?

Reply:
Yeah it should be closed here.

Adding this also revealed another bug, which I fixed in the same commit. Oddly, aborting write transactions seems to lead to memory corruption, which is detected as a double free when closing the env.

2b7c2da5a67994ff946d6ba1822f03bf3659a20c